tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php

```diff
@@ -349,22 +360,76 @@ public function test_config_printed_correctly_with_nested_empty_array() {
 
-		$this->assertEquals(
-			'{"config":{"myPlugin":{"tags":"Tags: \u003C!-- \u003Cscript\u003E"}},"state":{"myPlugin":{"amps":"http:\/\/site.test\/?foo=1\u0026baz=2"}}}',
-			$interactivity_data_string[1]
+		$expected = <<<"JSON"
+{"config":{"myPlugin":{"chars":"&\\u003C\\u003E/"}},"state":{"myPlugin":{"ampersand":"&","less-than sign":"\\u003C","greater-than sign":"\\u003E","solidus":"/","line separator":"\u{2028}","paragraph separator":"\u{2029}","flag of england":"\u{1F3F4}\u{E0067}\u{E0062}\u{E0065}\u{E006E}\u{E0067}\u{E007F}","malicious script closer":"\\u003C/script\\u003E","entity-encoded malicious script closer":"&lt;/script&gt;"}}}
+JSON;
+		$this->assertEquals( $expected, $interactivity_data_string[1] );
+	}
+
+	/**
+	 * Tests that special characters in the initial state and configuration are
+	 * properly escaped when the blog_charset is not UTF-8 (unicode compatible).
+	 *
+	 * This this test, unicode and line terminators should be escaped to their
+	 * JSON unicode sequences.
+	 *
+	 * @ticket 61170
+	 *
+	 * @covers ::state
+	 * @covers ::config
+	 * @covers ::print_client_interactivity_data
+	 */
+	public function test_state_and_config_escape_special_characters_non_utf8() {
+		add_filter( 'pre_option_blog_charset', array( $this, 'charset_iso_8859_1' ) );
+		$this->interactivity->state(
+			'myPlugin',
+			array(
+				'ampersand'                              => '&',
+				'less-than sign'                         => '<',
+				'greater-than sign'                      => '>',
+				'solidus'                                => '/',
+				'line separator'                         => "\u{2028}",
+				'paragraph separator'                    => "\u{2029}",
+				'flag of england'                        => "\u{1F3F4}\u{E0067}\u{E0062}\u{E0065}\u{E006E}\u{E0067}\u{E007F}",
+				'malicious script closer'                => '</script>',
+				'entity-encoded malicious script closer' => '&lt;/script&gt;',
+			)
 		);
+		$this->interactivity->config( 'myPlugin', array( 'chars' => '&<>/' ) );
+
+		$interactivity_data_markup = get_echo( array( $this->interactivity, 'print_client_interactivity_data' ) );
+		preg_match( '~<script type="application/json" id="wp-interactivity-data">\s*(\{.*\})\s*</script>~s', $interactivity_data_markup, $interactivity_data_string );
+
+		$expected = <<<"JSON"
+{"config":{"myPlugin":{"chars":"&\\u003C\\u003E/"}},"state":{"myPlugin":{"ampersand":"&","less-than sign":"\\u003C","greater-than sign":"\\u003E","solidus":"/","line separator":"\\u2028","paragraph separator":"\\u2029","flag of england":"\\ud83c\\udff4\\udb40\\udc67\\udb40\\udc62\\udb40\\udc65\\udb40\\udc6e\\udb40\\udc67\\udb40\\udc7f","malicious script closer":"\\u003C/script\\u003E","entity-encoded malicious script closer":"&lt;/script&gt;"}}}
+JSON;
+		$this->assertEquals( $expected, $interactivity_data_string[1] );
```

Comment:
Idea: This could use the HTML Processor to step over the tokens in `$interactivity_data_markup` as an additional check to ensure that the malicious script closer does not prematurely close the `script`. Maybe use the PHP's DOM API as well?

Reply:
I thought the same thing as I worked with these tests. However, in this case I think it makes sense to match and work with literal strings. I don't have much confidence in any of the parsers to do what I expect and not try to interpret any of this as HTML markup. I want to know exactly what characters are output and don't want any entities to be transformed.

Reply:
The HTML Processor doesn't support SCRIPT tags right now, and the tag processor is much more rudimentary. I'm not sure either is ready to handle what you suggest. @dmsnell thoughts on this?

Reply:
in this case the Tag Processor should be inscrutable for testing `SCRIPT` elements, because it does consume the entire thing in one go. you can compare `get_modifiable_text()` to your expectation to see what was inside the script